File: test/chebychev_test.cpp
// VolEsti (volume computation and sampling library)

// Copyright (c) 20012-2018 Vissarion Fisikopoulos
// Copyright (c) 2018 Apostolos Chalkis

// Licensed under GNU LGPL.3, see LICENCE file

#include "doctest.h"
#include <unistd.h>
#include "Eigen/Eigen"
#include <fstream>
#include "random.hpp"
#include "random/uniform_int.hpp"
#include "random/normal_distribution.hpp"
#include "random/uniform_real_distribution.hpp"
#include "volume.h"
#include "misc.h"
#include "known_polytope_generators.h"

template <typename NT>
NT factorial(NT n)
{
    return (n == 1 || n == 0) ? 1 : factorial(n - 1) * n;
}


template <typename NT, class RNGType, class Polytope>
void cheb_test(Polytope &P, NT expected, NT tolerance=0.0001)
{

    typedef typename Polytope::PolytopePoint Point;

    // Setup the parameters
    int n = P.dimension();
    int walk_len=10 + n/10;
    int nexp=1, n_threads=1;
    NT e=1, err=0.0000000001;
    int rnum = std::pow(e,-2) * 400 * n * std::log(n);
    RNGType rng(std::time(0));
    boost::normal_distribution<> rdist(0,1);
    boost::random::uniform_real_distribution<>(urdist);
    boost::random::uniform_real_distribution<> urdist1(-1,1);

    vars<NT,RNGType> var(rnum,n,walk_len,n_threads,err,e,0,0,0,0,0.0,rng,
             urdist,urdist1,-1.0,false,false,false,false,false,false,true, false,false);

    //Compute chebychev ball//
    //std::cout << "\n--- Testing Chebchev ball computation of " << f << std::endl;
    std::cout << "Number type: " << typeid(NT).name() << std::endl;
    double tstart1 = (double)clock()/(double)CLOCKS_PER_SEC;
    std::pair<Point,NT> CheBall = P.ComputeInnerBall();
    double tstop1 = (double)clock()/(double)CLOCKS_PER_SEC;


    std::cout<<"\nradius is: "<<CheBall.second<<std::endl;
    std::cout << "Chebychev time = " << tstop1 - tstart1 << std::endl;

    NT error = std::abs(CheBall.second-expected)/expected;

    CHECK(error < tolerance);

}

template <typename NT>
void call_test_cube() {
    typedef Cartesian<NT>    Kernel;
    typedef typename Kernel::Point    Point;
    typedef boost::mt19937    RNGType;
    typedef HPolytope<Point> Hpolytope;
    Hpolytope P;

    std::cout << "\n--- Testing Chebchev ball computation of H-cube10" << std::endl;
    P = gen_cube<Hpolytope>(10, false);
    cheb_test<NT, RNGType>(P, 1.0);

    std::cout << "\n--- Testing Chebchev ball computation of H-cube20" << std::endl;
    P = gen_cube<Hpolytope>(20, false);
    cheb_test<NT, RNGType>(P, 1.0);

    std::cout << "\n--- Testing Chebchev ball computation of H-cube30" << std::endl;
    P = gen_cube<Hpolytope>(30, false);
    cheb_test<NT, RNGType>(P, 1.0);

}

template <typename NT>
void call_test_cross() {
    typedef Cartesian<NT>    Kernel;
    typedef typename Kernel::Point    Point;
    typedef boost::mt19937    RNGType;
    typedef HPolytope<Point> Hpolytope;

    std::cout << "\n--- Testing Chebchev ball computation of H-cross10" << std::endl;
    Hpolytope P = gen_cross<Hpolytope>(10, false);
    cheb_test<NT, RNGType>(P, 0.316228);

}

template <typename NT>
void call_test_birk() {
    typedef Cartesian<NT>    Kernel;
    typedef typename Kernel::Point    Point;
    typedef boost::mt19937    RNGType;
    typedef HPolytope<Point> Hpolytope;
    Hpolytope P;

    std::cout << "\n--- Testing Chebchev ball computation of H-birk3" << std::endl;
    std::ifstream inp;
    std::vector<std::vector<NT> > Pin;
    inp.open("../R-proj/inst/extdata/birk3.ine",std::ifstream::in);
    read_pointset(inp,Pin);
    P.init(Pin);
    cheb_test<NT, RNGType>(P, 0.207107);

    std::cout << "\n--- Testing Chebchev ball computation of H-birk4" << std::endl;
    std::ifstream inp2;
    std::vector<std::vector<NT> > Pin2;
    inp2.open("../R-proj/inst/extdata/birk4.ine",std::ifstream::in);
    read_pointset(inp2,Pin2);
    P.init(Pin2);
    cheb_test<NT, RNGType>(P, 0.122008);

    std::cout << "\n--- Testing Chebchev ball computation of H-birk5" << std::endl;
    std::ifstream inp3;
    std::vector<std::vector<NT> > Pin3;
    inp3.open("../R-proj/inst/extdata/birk5.ine",std::ifstream::in);
    read_pointset(inp3,Pin3);
    P.init(Pin3);
    cheb_test<NT, RNGType>(P, 0.0833333);

    std::cout << "\n--- Testing Chebchev ball computation of H-birk6" << std::endl;
    std::ifstream inp4;
    std::vector<std::vector<NT> > Pin4;
    inp4.open("../R-proj/inst/extdata/birk6.ine",std::ifstream::in);
    read_pointset(inp4,Pin4);
    P.init(Pin4);
    cheb_test<NT, RNGType>(P, 0.0618034);

}

template <typename NT>
void call_test_prod_simplex() {
    typedef Cartesian<NT>    Kernel;
    typedef typename Kernel::Point    Point;
    typedef boost::mt19937    RNGType;
    typedef HPolytope<Point> Hpolytope;
    Hpolytope P;

    std::cout << "\n--- Testing Chebchev ball computation of H-prod_simplex5" << std::endl;
    P = gen_prod_simplex<Hpolytope>(5);
    cheb_test<NT, RNGType>(P, 0.138197);

    std::cout << "\n--- Testing Chebchev ball computation of H-prod_simplex10" << std::endl;
    P = gen_prod_simplex<Hpolytope>(10);
    cheb_test<NT, RNGType>(P, 0.0759747);

    std::cout << "\n--- Testing Chebchev ball computation of H-prod_simplex15" << std::endl;
    P = gen_prod_simplex<Hpolytope>(15);
    cheb_test<NT, RNGType>(P, 0.0529858);

    std::cout << "\n--- Testing Chebchev ball computation of H-prod_simplex20" << std::endl;
    P = gen_prod_simplex<Hpolytope>(20);
    cheb_test<NT, RNGType>(P, 0.0408628);

}

template <typename NT>
void call_test_simplex() {
    typedef Cartesian<NT>    Kernel;
    typedef typename Kernel::Point    Point;
    typedef boost::mt19937    RNGType;
    typedef HPolytope<Point> Hpolytope;
    Hpolytope P;

    std::cout << "\n--- Testing Chebchev ball computation of H-simplex10" << std::endl;
    P = gen_simplex<Hpolytope>(10, false);
    cheb_test<NT, RNGType>(P, 0.0759747);

    std::cout << "\n--- Testing Chebchev ball computation of H-simplex20" << std::endl;
    P = gen_simplex<Hpolytope>(20, false);
    cheb_test<NT, RNGType>(P, 0.0408628);

    std::cout << "\n--- Testing Chebchev ball computation of H-simplex30" << std::endl;
    P = gen_simplex<Hpolytope>(30, false);
    cheb_test<NT, RNGType>(P, 0.0281871);

    std::cout << "\n--- Testing Chebchev ball computation of H-simplex40" << std::endl;
    P = gen_simplex<Hpolytope>(40, false);
    cheb_test<NT, RNGType>(P, 0.0215868);

    std::cout << "\n--- Testing Chebchev ball computation of H-simplex50" << std::endl;
    P = gen_simplex<Hpolytope>(50, false);
    cheb_test<NT, RNGType>(P, 0.017522);

}

template <typename NT>
void call_test_skinny_cube() {
    typedef Cartesian<NT>    Kernel;
    typedef typename Kernel::Point    Point;
    typedef boost::mt19937    RNGType;
    typedef HPolytope<Point> Hpolytope;
    Hpolytope P;

    std::cout << "\n--- Testing Chebchev ball computation of H-skinny_cube10" << std::endl;
    P = gen_skinny_cube<Hpolytope>(10);
    cheb_test<NT, RNGType>(P, 1.0);

    std::cout << "\n--- Testing Chebchev ball computation of H-skinny_cube20" << std::endl;
    P = gen_skinny_cube<Hpolytope>(20);
    cheb_test<NT, RNGType>(P, 1.0);

}



TEST_CASE("cheb_cube") {
    call_test_cube<double>();
    call_test_cube<float>();
    call_test_cube<long double>();
}

TEST_CASE("cheb_cross") {
    call_test_cross<double>();
    call_test_cross<float>();
    call_test_cross<long double>();
}

TEST_CASE("cheb_birk") {
    call_test_birk<double>();
    call_test_birk<float>();
    call_test_birk<long double>();
}

TEST_CASE("cheb_prod_simplex") {
    call_test_prod_simplex<double>();
    call_test_prod_simplex<float>();
    call_test_prod_simplex<long double>();
}

TEST_CASE("cheb_simplex") {
    call_test_simplex<double>();
    call_test_simplex<float>();
    call_test_simplex<long double>();
}

TEST_CASE("cheb_skinny_cube") {
    call_test_skinny_cube<double>();
    call_test_skinny_cube<float>();
    call_test_skinny_cube<long double>();
}

File: cran_gen/cran_tests/chebychev_test.R
runCheTest <- function(A, b, name_string, radius, tol) {
  dimension = dim(A)[2]
  vec_ball = CheBall(A,b)
  rad = vec_ball[dimension + 1]
  
  error = abs(radius - rad) / radius
  print(paste0('error = ',error))
  if (error >= tol){
    print(paste0('TEST FAILED!! ', error, ' > ', tol))
  } else {
    print(paste0('Test PASSED!! [', name_string, ']'))
  }
  cat('\n')
}

library(volesti)

path = system.file('extdata', package = 'volesti')
tol = 0.00001

print('----------------------------------------')
print('------------1st test [cubes]------------')
print('----------------------------------------')
cat('\n')
print('H-cube10')
PolyList = GenCube(10, 'H')
runCheTest(PolyList$A, PolyList$b, 'H-cube10', 1.0, tol)

print('H-cube20')
PolyList = GenCube(20, 'H')
runCheTest(PolyList$A, PolyList$b, 'H-cube20', 1.0, tol)

print('H-cube30')
PolyList = GenCube(30, 'H')
runCheTest(PolyList$A, PolyList$b, 'H-cube30', 1.0, tol)

print('----------------------------------------')
print('------2nd test [cross_polytopes]--------')
print('----------------------------------------')
cat('\n')

print('H-cross10')
PolyList = GenCross(10, 'H')
runCheTest(PolyList$A, PolyList$b, 'H-cross10', 0.316228, tol)

print('----------------------------------------')
print('----------3rd test [birkhoff]-----------')
print('----------------------------------------')
cat('\n')
print('H-birk3')
ListPoly = fileToMatrix(paste0(path,'/birk3.ine'))
runCheTest(ListPoly$A, ListPoly$b, 'H-birk3', 0.207107, tol)

print('H-birk4')
ListPoly = fileToMatrix(paste0(path,'/birk4.ine'))
runCheTest(ListPoly$A, ListPoly$b, 'H-birk4', 0.122008, tol)

print('H-birk5')
ListPoly = fileToMatrix(paste0(path,'/birk5.ine'))
runCheTest(ListPoly$A, ListPoly$b, 'H-birk5', 0.0833333, tol)

print('H-birk6')
ListPoly = fileToMatrix(paste0(path,'/birk6.ine'))
runCheTest(ListPoly$A, ListPoly$b, 'H-birk6', 0.0618034, tol)

print('----------------------------------------')
print('--------4th test [prod_simplex]---------')
print('----------------------------------------')
cat('\n')
print('H-prod_simplex_5_5')
PolyList = GenProdSimplex(5)
runCheTest(PolyList$A, PolyList$b, 'H-prod_simplex_5_5', 0.138197, tol)

print('H-prod_simplex_10_10')
PolyList = GenProdSimplex(10)
runCheTest(PolyList$A, PolyList$b, 'H-prod_simplex_10_10', 0.0759747, tol)

print('H-prod_simplex_15_15')
PolyList = GenProdSimplex(15)
runCheTest(PolyList$A, PolyList$b, 'H-prod_simplex_15_15', 0.0529858, tol)

print('H-prod_simplex_20_20')
PolyList = GenProdSimplex(20)
runCheTest(PolyList$A, PolyList$b, 'H-prod_simplex_20_20', 0.0408628, tol)

print('----------------------------------------')
print('--------5th test [simplex]---------')
print('----------------------------------------')
cat('\n')
print('H-simplex10')
PolyList = GenSimplex(10, 'H')
runCheTest(PolyList$A, PolyList$b, 'H-simplex10', 0.0759747, tol)

print('H-simplex20')
PolyList = GenSimplex(20, 'H')
runCheTest(PolyList$A, PolyList$b, 'H-simplex20', 0.0408628, tol)

print('H-simplex30')
PolyList = GenSimplex(30, 'H')
runCheTest(PolyList$A, PolyList$b, 'H-simplex30', 0.0281871, tol)

print('H-simplex40')
PolyList = GenSimplex(40, 'H')
runCheTest(PolyList$A, PolyList$b, 'H-simplex40', 0.0215868, tol)

print('H-simplex50')
PolyList = GenSimplex(50, 'H')
runCheTest(PolyList$A, PolyList$b, 'H-simplex50', 0.017522, tol)

print('----------------------------------------')
print('--------6th test [skinny_cubes]---------')
print('----------------------------------------')
cat('\n')
print('H-skinny_cube10')
PolyList = GenSkinnyCube(10)
runCheTest(PolyList$A, PolyList$b, 'H-skinny_cube10', 1.0, tol)

print('H-skinny_cube20')
PolyList = GenSkinnyCube(20)
runCheTest(PolyList$A, PolyList$b, 'H-skinny_cube20', 1.0, tol)
